drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,293 @@
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_CMS) || (return_code == AT_CMD_CME)) {
+		last_error = atoi(&buf[ARRAY_SIZE(AT_CMD_CMS_STR) - 1]);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	if ((data != NULL) && (data->callback != NULL)) {
+		data->callback(data->data);
+	}
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("AT socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+		write_with_response = false;
+
+		if (bytes_read < 0) {
+			LOG_ERR("AT socket recv failed with err %d",
+				bytes_read);
+
+			if (errno == -EAGAIN) {
+				if (open_socket() == 0) {
+					return_code = AT_CMD_ERROR;
+					goto next;
+				}
+			}
+
+			LOG_ERR("Unrecoverable reception error (err: %d),"
+				"thread killed", errno);
+			return;
+		} else if (bytes_read == sizeof(buf) ||
+			   buf[bytes_read - 1] != '\0') {
+
+			LOG_ERR("AT message to large for reception buffer or "
+				"missing termination character");
+			return_code = AT_CMD_ERROR;
+			goto next;
+		}
+
+		return_code = get_return_code(buf);
+
+		if (return_code != AT_CMD_NOTIFICATION) {
+			if ((response_buf_len > 0) &&
+			    (response_buf != NULL)) {
+				if (response_buf_len > strlen(buf)) {
+					strncpy(response_buf,
+						buf,
+						response_buf_len);
+				} else {
+					LOG_ERR("Response buffer not large "
+						"enough");
+					response_buf = NULL;
+				}
+
+				write_with_response = true;
+				response_buf_len    = 0;
+				response_buf        = NULL;
+			}
+		}
+
+		if (strlen(buf) == 0) {
+			goto next;
+		}
+
+		if (!write_with_response) {
+			struct callback_work_item *item =
+			k_malloc(sizeof(struct callback_work_item));
+
+			if (!item) {
+				LOG_DBG("Failed to allocate work item");
+				goto next;
+			}
+
+			k_work_init(&item->work, callback_worker);
+
+			memcpy(item->data, buf, strlen(buf) + 1);
+
+			if (return_code == AT_CMD_NOTIFICATION) {
+				item->callback = notification_handler;
+			} else {
+				item->callback = current_cmd_handler;
+			}
+
+			k_work_submit(&item->work);
+		}
+next:
+		if (return_code != AT_CMD_NOTIFICATION) {
+			k_msgq_put(&return_code_msq, &return_code, K_FOREVER);
+		}
+		continue;
+	}
+}
+
+static inline enum at_cmd_rc at_write(const char *const cmd)
```

Comment:
I do not think this needs to be inlined.

Reply:
I want to give the compiler and the reader a hint about my intention here. I want this code to be a part the caller and not a standalone function for effency and probably better optimization. This is ofcourse something that the compiler probably would have done anyway, but the inline makes it a wanted behaviour from my side.